### PR TITLE
Basic Initial SSH implementation

### PR DESCRIPTION
**Pull request by ygins**
- Server with customizable key retrieval methods - configured by passing a DataStore (either local or S3) into it.
- Unit tests for publisher/server - local and remote
- For testing, set bucket name via env variable SSH_KEY_DEV_BUCKET_NAME
- before testing, run the dev-bootstrap script located in tools/dev-bootstrap.sh. This currently overpermissions, will figure out which ones are needed or not and update accordingly
- To test, from the root directory run python3.11 -m unittest

### CODE REVIEW

File: ssh_key_rotator/publisher/key_publisher.py [hunk 1]
```diff
@@ -0,0 +1,85 @@
...
+        # Store the new public key in S3 bucket
+        s3_client = boto3.client("s3")
+        s3_client.put_object(
+            Body=public_key,
+            Bucket=self.bucket_name,
+            Key=f"{self.host}/{self.user_id}-cert.pub",
+        )
+
+        # Store the private key on the local machine
+        private_key_path = f"{get_user_path()}/.ssh/{self.host}/{self.user_id}"
```
Comment (thehcma) [inline ssh_key_rotator/publisher/key_publisher.py:42]: we should be using a temp location

File: ssh_key_rotator/publisher/key_publisher.py [hunk 2]
```diff
@@ -0,0 +1,85 @@
...
+    server_flag_index = command_line_arguments.index("--server")
+    server = command_line_arguments[server_flag_index + 1]
+    data_store_flag_index = command_line_arguments.index("--datastore")
+    data_store = command_line_arguments[data_store_flag_index + 1]
+
+    if data_store == "local":
+        publisher = LocalPublisher(server, username)
+        publisher.publish_new_key()
+    elif data_store == "s3":
+        s3_bucket_name = input("Enter a bucket name: ")
```
Comment (thehcma) [inline ssh_key_rotator/publisher/key_publisher.py:81]: the bucket name should be a command line argument

File: ssh_key_rotator/publisher/key_publisher.py [hunk 3]
```diff
@@ -0,0 +1,85 @@
...
+        _, public_key = generate_private_public_key_in_file(
+            f"{user_path}/.ssh/{self.host}",
+            private_key_name=self.user_id,
+            public_key_name=f"{self.user_id}-cert.pub",
+        )
+        return public_key.decode()
+
+
+if __name__ == "__main__":
+    command_line_arguments = sys.argv
```
Comment (thehcma) [inline ssh_key_rotator/publisher/key_publisher.py:69]: use `argparse` for command-line argument parsing

File: ssh_key_rotator/publisher/key_publisher.py [hunk 4]
```diff
@@ -0,0 +1,85 @@
...
+
+
+class LocalPublisher(Publisher):
+    """Local Publisher class"""
+
+    def __init__(self, host: str, user_id: str):
+        self.host = host
+        self.user_id = user_id
+
+    def publish_new_key(self) -> str:
```
Comment (thehcma) [inline ssh_key_rotator/publisher/key_publisher.py:58]: what happens with the private key?

File: tests/unit/server/test_local.py [hunk 5]
```diff
@@ -0,0 +1,37 @@
...
+            key_fingerprint = key.fingerprint  # type: ignore
+            client = SSHClient()
+            client.load_system_host_keys()
+            client.set_missing_host_key_policy(AutoAddPolicy())
+            client.connect(
+                "127.0.0.1",
+                port=server.port,
+                pkey=key,
+            )
+            self.assertTrue(any(key_fingerprint in line for line in await server.logs))
```
Comment (thehcma) [inline tests/unit/server/test_local.py:37]: let's use `assert_that` from `pyhamcrest`

File: tests/unit/server/test_local.py [hunk 6]
```diff
@@ -0,0 +1,37 @@
...
+            private_key, _ = generate_private_public_key_in_file(
+                random_host_dir,
+                private_key_name=get_username(),
+                public_key_name=f"{get_username()}-cert.pub",
+            )
+            key: RSAKey = RSAKey.from_private_key(StringIO(private_key.decode()))
+            key_fingerprint = key.fingerprint  # type: ignore
+            client = SSHClient()
+            client.load_system_host_keys()
+            client.set_missing_host_key_policy(AutoAddPolicy())
```
Comment (thehcma) [inline tests/unit/server/test_local.py:31]: this is usually unsafe - you should be using
```
ssh_client.set_missing_host_key_policy(RejectPolicy())
```

and then explicitly add the correct host key (automate this - find out what the host key is and add it), something like:

ssh_client.get_host_keys().add(....)

if you are ambitious you should be emitting the host key yourself and adding it as part of your sshd config

File: tests/unit/server/test_remote.py [hunk 7]
```diff
@@ -0,0 +1,45 @@
...
+            ssh_client.load_system_host_keys()
+            ssh_client.set_missing_host_key_policy(AutoAddPolicy())
+            ssh_client.connect(
+                "127.0.0.1",
+                server.port,
+                username=get_username(),
+                pkey=private_key_paramiko,
+            )
+            key_fingerprint = private_key_paramiko.fingerprint  # type: ignore
+            self.assertTrue(any(key_fingerprint in line for line in await server.logs))
```
Comment (thehcma) [inline tests/unit/server/test_remote.py:45]: assert_that

File: tools/dev-bootstrap.sh [hunk 8]
```diff
@@ -0,0 +1,24 @@
+#!/bin/bash
+
+USERNAME="$(whoami)"
+SUDOERS_FILE="/etc/sudoers.d/$USERNAME"
+
+# Add the user to the sudo group if not already a member
+sudo usermod -aG sudo "$USERNAME"
+
+# Overwrite the sudoers configuration
+echo "$USERNAME ALL=(ALL) NOPASSWD: /bin/chmod, /bin/chgrp, /usr/bin/chmod, /usr/bin/chgrp, /bin/cp, /usr/bin/cp, /bin/touch, /usr/bin/touch, /bin/mkdir, /usr/bin/mkdir" | sudo tee "$SUDOERS_FILE" >/dev/null
```
Comment (thehcma) [inline tools/dev-bootstrap.sh:10]: ideally this should be one single command (the `rsync` I suggested)

**Review by thehcma — APPROVED** [review]
This is shaping up very well, but there are more improvements I'd like to see.

Perhaps in a subsequent PR.

**Comment by ygins** [comment]
Thank you, we will look into these changes. Merging to main

File: ssh_key_rotator/publisher/key_publisher.py (hunk 2 above)
Comment (jesselanger2) [reply ssh_key_rotator/publisher/key_publisher.py:81]: That's what we're doing with input. If the user writes --datastore s3 we will ask him for the bucket name.

Comment (ygins) [reply ssh_key_rotator/publisher/key_publisher.py:81]: He means directly input with the command, like the other options

File: ssh_key_rotator/custom_keygen.py [hunk 9]
```diff
@@ -1,14 +1,30 @@
 "Module for generating public/private SSH key pairs"
 import os
 import shutil
+from typing import Tuple
```
[the remark this answers is not in the document]
Comment (ygins) [reply ssh_key_rotator/custom_keygen.py:4]: On it

File: ssh_key_rotator/custom_keygen.py [hunk 10]
```diff
@@ -1,14 +1,30 @@
...
 
+
 PRIVATE_KEY_NAME: str = "key"
 PUBLIC_KEY_NAME: str = f"{PRIVATE_KEY_NAME}-cert.pub"
 
-def generate_private_public_key_in_file(public_key_dir: str, private_key_dir: str|None = None):
+
+def generate_private_public_key() -> Tuple[bytes, bytes]:
+    "Generates a private and public RSA key"
+    key = RSA.generate(1024)
```
[the remark this answers is not in the document]
Comment (ygins) [reply ssh_key_rotator/custom_keygen.py:14]: on it

File: ssh_key_rotator/custom_keygen.py [hunk 11]
```diff
@@ -1,14 +1,30 @@
...
+    private_key = key.export_key()
+    public_key = key.public_key().export_key(format="OpenSSH")
+    return private_key, public_key
+
+
+def generate_private_public_key_in_file(
```
[the remark this answers is not in the document]
Comment (ygins) [reply ssh_key_rotator/custom_keygen.py:20]: In our code, we have the Publisher as a separate entity from the DataStore - this is just a utility method to generate keys. Can you clarify what you mean by this should be using the DataStore API's?

File: ssh_key_rotator/publisher/key_publisher.py [hunk 12]
```diff
@@ -0,0 +1,85 @@
...
+    @abstractmethod
+    def publish_new_key(self) -> str:
+        """Abstract method for publishing a new public key"""
+
+
+class S3Publisher(Publisher):
+    """S3 Publisher class"""
+
+    def __init__(self, bucket_name: str, host: str, user_id: str):
+        self.bucket_name = bucket_name
```
[the remark this answers is not in the document]
Comment (ygins) [reply ssh_key_rotator/publisher/key_publisher.py:25]: Yes but I thought __init__ is specifically different

[the remark this answers is not in the document]
Comment (ygins) [reply ssh_key_rotator/publisher/key_publisher.py:25]: * __ init __

File: ssh_key_rotator/publisher/key_publisher.py (hunk 1 above)
Comment (ygins) [reply ssh_key_rotator/publisher/key_publisher.py:42]: We will include that functionality and allow for a different directory to be passed in. The DataStore already takes care of the temporary bit

File: ssh_key_rotator/publisher/key_publisher.py (hunk 4 above)
Comment (ygins) [reply ssh_key_rotator/publisher/key_publisher.py:58]: Gets stored on the host machine via the generate method

File: tests/unit/server/test_local.py (hunk 5 above)
Comment (ygins) [reply tests/unit/server/test_local.py:37]: on it

File: tests/unit/server/test_remote.py (hunk 7 above)
Comment (ygins) [reply tests/unit/server/test_remote.py:45]: on it

File: ssh_key_rotator/custom_keygen.py (hunk 10 above)
[the remark this answers is not in the document]
Comment (ygins) [reply ssh_key_rotator/custom_keygen.py:14]: next PR

File: tests/unit/server/test_local.py (hunk 5 above)
Comment (ygins) [reply tests/unit/server/test_local.py:37]: next PR

File: tests/unit/server/test_remote.py (hunk 7 above)
Comment (ygins) [reply tests/unit/server/test_remote.py:45]: next PR

**Pull request merged**